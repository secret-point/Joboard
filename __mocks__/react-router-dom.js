// Mock navigate and useParams when needed
// const navigate = jest.fn();
// const useParams = jest.fn();
const useLocation = jest.fn();

module.exports = {
  ...jest.requireActual("react-router-dom"),
  // useNavigate: () => navigate,
  // useParams,
  useLocation,
};
