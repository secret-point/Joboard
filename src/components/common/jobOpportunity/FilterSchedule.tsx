import React from "react";
import { Col, Row } from "@amzn/stencil-react-components/layout";
import { Label, Text } from "@amzn/stencil-react-components/text";
import { Select } from "@amzn/stencil-react-components/form";
import { translate as t } from "../../../utils/translator";
import { DayHoursFilter, ScheduleStateFilters } from "../../../utils/types/common";
import DaysHoursFilter from "../daysHoursFilter/DaysHoursFilter";
import { DESIRED_WORK_HOURS } from "../../../utils/enums/common";
import { boundUpdateScheduleFilters } from "../../../actions/ScheduleActions/boundScheduleActions";

interface FilterScheduleProps {
    filters: ScheduleStateFilters
}

const FilterSchedule = ( props: FilterScheduleProps ) => {

    const { filters } = props;
    const { daysHoursFilter, maxHoursPerWeek } = filters;

    const updateMaxHoursPerWeek = ( key: DESIRED_WORK_HOURS ) => {
        const newFilters = { ...filters };
        newFilters.maxHoursPerWeek = key;
        boundUpdateScheduleFilters(newFilters);
    }

    const updateDaysHourFilter = ( dayHourFilters: DayHoursFilter[] ) => {
        const newFilters = { ...filters };
        newFilters.daysHoursFilter = dayHourFilters;
        boundUpdateScheduleFilters(newFilters);
    }

    return (
        <Col gridGap={20}>
            <Col gridGap={10} width="100%" padding={{ right: 'S300' }}>
                <Label htmlFor='desiredHoursSelect'>
                    {t("BB-FilterSchedule-desired-hour-select-label-text", "Desired hours per week")}
                </Label>
                <Select
                    id="desiredHoursSelect"
                    onChange={( option ) => updateMaxHoursPerWeek(option)}
                    options={Object.values(DESIRED_WORK_HOURS)}
                    renderOption={option => (
                        <Row width='100%'>
                            <Text>{t('BB-FilterSchedule-DesiredHours-Upto-hours', "Upto {hourNumber} hours", { hourNumber: option })}</Text>
                        </Row>
                    )}
                    defaultValue={maxHoursPerWeek}
                    renderNativeOption={option => t('BB-FilterSchedule-DesiredHours-Upto-hours', "Upto {hourNumber} hours", { hourNumber: option })}
                    renderSelectedValue={option => t('BB-FilterSchedule-DesiredHours-Upto-hours', "Upto {hourNumber} hours", { hourNumber: option })}
                    valueAccessor={option => option}
                    placeholder={t("BB-FilterSchedule-desired-hour-select-place-holder", "Select Desired Hours")}
                />
            </Col>
            <DaysHoursFilter
                defaultFilter={daysHoursFilter}
                label={t("BB-DaysHoursFilter-title-text", "Filter Schedules")}
                onValueChange={updateDaysHourFilter}
            />
        </Col>
    )
}

export default FilterSchedule;
