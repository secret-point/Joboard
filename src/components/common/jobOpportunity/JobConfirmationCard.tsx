import React from "react";
import { Col, Row } from "@amzn/stencil-react-components/layout";
import { Text } from "@amzn/stencil-react-components/text";
import {
    IconArrowRight,
    IconCalendarFill,
    IconClockFill,
    IconDocument,
    IconGlobe,
    IconPaymentFill,
    IconSize
} from '@amzn/stencil-react-components/icons';
import { Link } from "@amzn/stencil-react-components/link";
import VideoContainer from "../VideoContainer";
import { boundSetJobOpportunityPage } from "../../../actions/UiActions/boundUi";
import { JOB_OPPORTUNITY_PAGE } from "../../../utils/enums/common";
import { Schedule } from "../../../utils/types/common";
import { translate as t } from "../../../utils/translator";

interface JobConfirmationCardProps {
    schedule: Schedule
}

const JobConfirmationCard = ( props: JobConfirmationCardProps ) => {
    const { schedule } = props;

    const {
        hoursPerWeek,
        firstDayOnSite,
        currencyCode,
        postalCode,
        city,
        state,
        scheduleText,
        address,
        totalPayRate,
        jobPreviewVideo,
        scheduleBannerText
    } = schedule;

    return (
        <Col className="jobConfirmationCardContainer">
            <Row className="jobConfirmationHeader" gridGap={12}>
                <VideoContainer
                    id='jobVideo'
                    src={jobPreviewVideo}/>
            </Row>
            <Row padding="S300" alignItems="center">
                <Col gridGap={8} width="95%">
                    <Row gridGap={5} alignItems="center">
                        <IconClockFill size={IconSize.ExtraSmall}/>
                        <Text fontSize="T100">{scheduleText}</Text>
                    </Row>
                    <Row gridGap={5} alignItems="center">
                        <IconCalendarFill size={IconSize.ExtraSmall}/>
                        <Text fontSize="T100">{`Possible Start Date: ${firstDayOnSite}`} <b>(or earlier!)</b></Text>
                    </Row>
                    <Row gridGap={5} alignItems="center">
                        <IconPaymentFill size={IconSize.ExtraSmall}/>
                        <Text fontSize="T100">{`${currencyCode}${totalPayRate} /hr`}</Text>
                    </Row>
                    <Row gridGap={5} alignItems="center">
                        <IconDocument size={IconSize.ExtraSmall}/>
                        <Text fontSize="T100">{`${hoursPerWeek}hrs/wk`}</Text>
                    </Row>
                    <Row gridGap={5} alignItems="center">
                        <IconGlobe size={IconSize.ExtraSmall}/>
                        <Text fontSize="T100">{`${address}, ${city}, ${state} ${postalCode}`}</Text>
                    </Row>
                </Col>
            </Row>
            {
                scheduleBannerText &&
                <Row padding={{ left: 'S300', right: 'S300', bottom: 'S300' }}>
                    <Link fontSize='T100'>{scheduleBannerText}</Link>
                </Row>
            }
            <Row
                gridGap={8}
                alignItems="center"
                padding={'S200'}
                className='jobDetailLink'
                onClick={() => {
                    boundSetJobOpportunityPage(JOB_OPPORTUNITY_PAGE.JOB_DESCRIPTION);
                }}
            >
                <Text fontWeight="medium">
                    {t('BB-JobOpportunity-go-to-jobDescription-link', 'Job requirement and benefits')}
                </Text>
                <IconArrowRight size={IconSize.ExtraSmall}/>
            </Row>
        </Col>
    )
}

export default JobConfirmationCard;
