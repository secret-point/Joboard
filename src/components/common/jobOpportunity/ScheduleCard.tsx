import {
    IconChevronRight
} from '@amzn/stencil-react-components/icons';
import { Col, Row } from "@amzn/stencil-react-components/layout";
import { Link } from "@amzn/stencil-react-components/link";
import { Text } from "@amzn/stencil-react-components/text";
import React from "react";
import { CommonColors } from "../../../utils/colors";
import { QUERY_PARAMETER_NAME } from "../../../utils/enums/common";
import { routeToAppPageWithPath } from "../../../utils/helper";
import { QueryParamItem, Schedule } from "../../../utils/types/common";
import { PAGE_ROUTES } from "../../pageRoutes";
import ScheduleCardBanner from "./ScheduleCardBanner";
import ScheduleDetails from "./ScheduleDetails";
import Image from "../../common/Image";

interface ScheduleCardProps {
    scheduleDetail: Schedule,
    displayOnly?: boolean
}

const ScheduleCard = ( props: ScheduleCardProps ) => {

    const { scheduleDetail, displayOnly } = props;
    const { JOB_CONFIRMATION } = PAGE_ROUTES;

    const {
        image,
        bonusSchedule,
        signOnBonus,
        signOnBonusL10N,
        currencyCode,
        tagLine,
        externalJobTitle,
        scheduleBannerText
    } = scheduleDetail;

    const handleClick = () => {
        if(!displayOnly) {
            const queryParamItem: QueryParamItem = {
                paramName: QUERY_PARAMETER_NAME.SCHEDULE_ID,
                paramValue: scheduleDetail.scheduleId
            }
            routeToAppPageWithPath(JOB_CONFIRMATION, [queryParamItem]);
        }
    }

    return (
        <Col
            className="scheduleCardContainer"
            padding={{ bottom: 'S300' }}
            onClick={handleClick}
        >
            <Row className="scheduleCardHeader" gridGap={12} padding="S300" backgroundColor={CommonColors.Neutral10}>
                <Col className="scheduleCardBannerContainer" gridGap={4}>
                    <Image
                        src={image}/>
                    {
                        bonusSchedule &&
                        <ScheduleCardBanner signOnBonusL10N={signOnBonusL10N} currencyCode={currencyCode} signOnBonus={signOnBonus}/>
                    }
                </Col>
                <Col width="auto" gridGap={10}>
                    <Text fontSize='T200'>{externalJobTitle}</Text>
                    <Text fontSize='T100'>{tagLine}</Text>
                </Col>
            </Row>
            <Row padding="S300" alignItems="center">
                <ScheduleDetails scheduleDetail={scheduleDetail} />
                {!displayOnly &&
                <Col>
                    <IconChevronRight/>
                </Col>}
            </Row>
            {
                scheduleBannerText &&
                <Row
                  padding={{ left: 'S300', right: 'S300', bottom: 'S300' }}
                  className="scheduleBannerText"
                >
                    <Text fontSize='T100'>{scheduleBannerText}</Text>
                </Row>
            }
        </Col>
    )
}

export default ScheduleCard;
