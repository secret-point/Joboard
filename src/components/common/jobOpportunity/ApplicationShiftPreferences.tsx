import React, { useEffect, useState } from "react";
import { Col, Row } from "@amzn/stencil-react-components/layout";
import { H5, Text } from "@amzn/stencil-react-components/text";
import { translate as t } from "../../../utils/translator";
import { IconArrowLeft, IconSize } from "@amzn/stencil-react-components/icons";
import { PopupDatePicker } from "@amzn/stencil-react-components/date-picker";
import {
  checkAndBoundGetApplication,
  getLocale,
  isDateGreaterThanToday,
  routeToAppPageWithPath,
  UpdateHoursPerWeekHelper
} from "../../../utils/helper";
import { Checkbox, InputFooter, InputWrapper, LabelPosition, Radio } from "@amzn/stencil-react-components/form";
import { CommonColors } from "../../../utils/colors";
import {
  shiftPreferenceShiftPattern,
  shiftPreferenceApplicationWeekDays,
  shiftPreferenceApplicationWeekendDays,
  shiftPreferenceWorkHour
} from "../../../utils/constants/common";
import { Button, ButtonVariant } from "@amzn/stencil-react-components/button";
import { PAGE_ROUTES } from "../../pageRoutes";
import { Application, ShiftPreferenceWorkHour } from "../../../utils/types/common";
import { JobState } from "../../../reducers/job.reducer";
import { ApplicationState } from "../../../reducers/application.reducer";
import { CandidateState } from "../../../reducers/candidate.reducer";
import { useLocation } from "react-router-dom";
import {
  getPageNameFromPath,
  parseQueryParamsArrayToSingleItem,
  resetIsPageMetricsUpdated
} from "../../../helpers/utils";
import queryString from "query-string";
import { boundGetCandidateInfo } from "../../../actions/CandidateActions/boundCandidateActions";
import { boundGetJobDetail } from "../../../actions/JobActions/boundJobDetailActions";
import { addMetricForPageLoad } from "../../../actions/AdobeActions/adobeActions";
import { DAYS_OF_WEEK, UPDATE_APPLICATION_API_TYPE, WORKFLOW_STEP_NAME } from "../../../utils/enums/common";
import { UpdateApplicationRequestDS } from "../../../utils/apiTypes";
import { connect } from "react-redux";
import moment from "moment";
import { boundUpdateApplicationDS } from "../../../actions/ApplicationActions/boundApplicationActions";
import { onCompleteTaskHelper } from "../../../actions/WorkflowActions/workflowActions";

interface MapStateToProps {
  job: JobState;
  application: ApplicationState;
  candidate: CandidateState;
}

export const ApplicationShiftPreferences = (props: MapStateToProps) => {

  const { job, application, candidate } = props;
  const { search, pathname } = useLocation();
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  const pageName = getPageNameFromPath(pathname);
  const queryParams = parseQueryParamsArrayToSingleItem(queryString.parse(search));
  const { applicationId, jobId } = queryParams;
  const jobDetail = job.results;
  const applicationData = application.results;
  const { candidateData } = candidate.results;

  useEffect(() => {
    boundGetCandidateInfo();
  }, []);

  useEffect(() => {
    jobId && jobId !== jobDetail?.jobId && boundGetJobDetail({ jobId: jobId, locale: getLocale() });
  }, [jobDetail, jobId]);

  useEffect(() => {
    checkAndBoundGetApplication(applicationId);
  }, [applicationId]);

  useEffect(() => {
    // Page will emit page load event once both pros are available but
    // will not emit new event on props change once it has emitted pageload event previously
    jobDetail && applicationData && candidateData && addMetricForPageLoad(pageName);

  }, [jobDetail, applicationData, candidateData]);

  useEffect(() => {
    return () => {
      // reset this so as it can emit new pageload event after being unmounted.
      resetIsPageMetricsUpdated(pageName);
    };
  }, []);

  const [startDate, setStartDate] = useState();
  const [hoursPerWeek, setHoursPerWeek] = useState<ShiftPreferenceWorkHour[]>([]);
  const [workDays, setWorkDays] = useState<DAYS_OF_WEEK[]>([]);
  const [shiftPattern, setShiftPattern] = useState();
  const [startDateValid, setStartDateValid] = useState(true);
  const [shiftPatternValid, setShiftPatternValid] = useState(true);
  const [hourPerWeekValid, setHourPerWeekValid] = useState(true);
  const [workDaysValid, setWorkDaysValid] = useState(true);

  const updateHoursPerWeek = (value: ShiftPreferenceWorkHour) => {
    setHoursPerWeek(UpdateHoursPerWeekHelper(hoursPerWeek || [], value));
  };

  const updateWorkDays = (value: DAYS_OF_WEEK) => {
    const newWorkDays = workDays || [];

    if (newWorkDays.indexOf(value) > -1) {
      newWorkDays.splice(newWorkDays.indexOf(value), 1);
    } else {
      newWorkDays.push(value);
    }

    setWorkDays(newWorkDays);
  };

  const handleSavePreference = () => {
    // Validate start date and shift pattern as they are required
    const isStartDateValid = isDateGreaterThanToday(startDate);
    const isShiftPatternValid = !!shiftPattern;
    const isHourPerWeekValid = hoursPerWeek.length > 0;
    const isWorkDaysValid = workDays.length > 0;
    const formIsValid = isShiftPatternValid && isWorkDaysValid && isHourPerWeekValid && isStartDateValid;

    setStartDateValid(isStartDateValid);
    setShiftPatternValid(isShiftPatternValid);
    setHourPerWeekValid(isHourPerWeekValid);
    setWorkDaysValid(isWorkDaysValid);

    if (formIsValid) {
      const request: UpdateApplicationRequestDS = {
        applicationId: applicationId,
        type: UPDATE_APPLICATION_API_TYPE.JOB_PREFERENCES,
        payload: {
          shiftPreference: {
            hoursPerWeek: hoursPerWeek,
            earliestStartDate: moment(startDate).format("DD/MM/YYYY"),
            jobRoles: [],
            shiftTimePattern: shiftPattern,
            candidateTimezone: candidateData?.timezone || "",
            daysOfWeek: workDays
          },
        }
      };

      boundUpdateApplicationDS(request, (applicationData: Application) => {
        onCompleteTaskHelper(applicationData, true, WORKFLOW_STEP_NAME.ADDITIONAL_INFORMATION);
      });
    }
  };

  return (
    <Col gridGap={10}>
      <Row padding={{ top: "S400" }}>
        <Row
          className="shiftPreferenceLink"
          gridGap={8}
          onClick={ () => {
            routeToAppPageWithPath(PAGE_ROUTES.JOB_OPPORTUNITIES);
          }}

        >
          <IconArrowLeft size={IconSize.Medium} fontSize="T100" aria-hidden />
          <Text fontWeight="medium" fontSize="T200">
            {t("BB-JobOpportunity-Go-To-shift-selection-Link", "Go back to shift selection")}
          </Text>
        </Row>
      </Row>
      <Col padding={{ top: "S300" }} gridGap={8}>
        <H5>
          {t("BB-shiftPreference-title-text", "Specify your job preferences")}
        </H5>
        <Col padding={{ top: "S200" }}>
          <InputWrapper
            id="startDateEnd"
            labelText={t("BB-ShiftPreference-start-date-label-text", "Select the earliest start date")}
            required
            error={!startDateValid}
            footer={!startDateValid ?
              t("BB-shiftPreference-start-date-error-text", "Please select a start date in the future") : undefined}
          >
            {inputProps => (
              <PopupDatePicker
                inputProps={{ ...inputProps, width: "100%" }}
                isDateDisabled={value => !isDateGreaterThanToday(value)}
                onChange={setStartDate}
              />
            )}
          </InputWrapper>
        </Col>
      </Col>
      <Col gridGap={10} padding={{ top: "S300" }}>
        <Text fontSize="T200" color={!hourPerWeekValid ? CommonColors.RED70 : "inherit"}>
          {t("BB-ShiftPreference-hours-per-week-label-text", "Select the number of hours that you are able to work per week")} *
        </Text>
        <Text fontSize="T100" color={CommonColors.Neutral70}>
          {t("BB-ShiftPreference-hours-per-week-label-text2", "Your visa may restrict the number of hours per week that you are legally authorised to work. If this is the case, choose the options that correspond to your visa restrictions.")}
        </Text>
        <Col gridGap={10} padding={{ top: "S300" }}>
          {
            shiftPreferenceWorkHour.map(workHour => (
              <InputWrapper
                id={`shiftWorkHour-${workHour.displayValue}`}
                labelText={t(workHour.translationKey, workHour.displayValue)}
                labelPosition={LabelPosition.Trailing}
                key={workHour.displayValue}
              >
                {inputProps => (
                  <Checkbox
                    name={workHour.displayValue}
                    {...inputProps}
                    onChange={() => updateHoursPerWeek(workHour.value)}
                  />
                )}
              </InputWrapper>
            ))
          }
        </Col>
        {!hourPerWeekValid && (
          <InputFooter error={!hourPerWeekValid} id="hoursPerWeekErrorContainer">
            {t("BB-ShiftPreference-hours-per-week-error-text", "Please select at least one option for the number of hours you are able to work per week.")}
          </InputFooter>
        )}
      </Col>
      <Col gridGap={10} padding={{ top: "S300" }}>
        <Text fontSize="T200" color={!workDaysValid ? CommonColors.RED70 : "inherit"}>
          {t("BB-ShiftPreference-workDays-label-text", "Select all the days of the week that you are able to work")} *
        </Text>
        <Text fontSize="T100" color={CommonColors.Neutral70}>
          {t("BB-ShiftPreference-workDays-label-text2", "Please note, most of our shift patterns will include at least one day of the weekend and shift preferences are not guaranteed.")}
        </Text>
        <Col padding={{ top: "S200" }} gridGap={10}>
          <Text fontSize="T100" color={CommonColors.Neutral70}>
            {t("BB-Weekends-text", "WEEKENDS")}
          </Text>
          <Col gridGap={10}>
            {
              shiftPreferenceApplicationWeekendDays.map(day => (
                <InputWrapper
                  id={`shiftWeekendDays-${day.displayValue}`}
                  labelText={t(day.translationKey, day.displayValue)}
                  labelPosition={LabelPosition.Trailing}
                  key={day.displayValue}
                >
                  {inputProps => (
                    <Checkbox
                      name={day.displayValue}
                      {...inputProps}
                      onChange={() => updateWorkDays(day.value as DAYS_OF_WEEK)}
                    />
                  )}
                </InputWrapper>
              ))
            }
          </Col>
        </Col>
        <Col padding={{ top: "S200" }} gridGap={10}>
          <Text fontSize="T100" color={CommonColors.Neutral70}>
            {t("BB-Weekdays-text", "WEEKDAYS")}
          </Text>
          <Col gridGap={10}>
            {
              shiftPreferenceApplicationWeekDays.map(day => (
                <InputWrapper
                  id={`shiftWeekdays-${day.displayValue}`}
                  labelText={t(day.translationKey, day.displayValue)}
                  labelPosition={LabelPosition.Trailing}
                  key={day.displayValue}
                >
                  {inputProps => (
                    <Checkbox
                      name={day.displayValue}
                      {...inputProps}
                      onChange={() => updateWorkDays(day.value as DAYS_OF_WEEK)}
                    />
                  )}
                </InputWrapper>
              ))
            }
          </Col>
        </Col>
        {!workDaysValid && (
          <InputFooter error={!workDaysValid} id="worDaysErrorContainer">
            {t("BB-ShiftPreference-wordDays-error-text", "Please select at least one day of the week that you are able to work.")}
          </InputFooter>
        )}
      </Col>
      <Col gridGap={10} padding={{ top: "S300" }}>
        <Text fontSize="T200" color={!shiftPatternValid ? CommonColors.RED70 : "inherit"}>
          {t("BB-ShiftPreference-shift-pattern-label-text", "Select the shift pattern that you are able to work")} *
        </Text>
        <Col gridGap={10}>
          {
            shiftPreferenceShiftPattern.map(shiftPattern => (
              <InputWrapper
                id={`shiftWeekendDays-${shiftPattern.displayValue}`}
                labelText={t(shiftPattern.translationKey, shiftPattern.displayValue)}
                labelPosition={LabelPosition.Trailing}
                key={shiftPattern.displayValue}
              >
                {inputProps => (
                  <Radio
                    name="shift-pattern"
                    {...inputProps}
                    onChange={() => setShiftPattern(shiftPattern.value)}
                  />
                )}
              </InputWrapper>
            ))
          }
          {!shiftPatternValid && (
            <InputFooter error={!shiftPatternValid} id="shiftPatternErrorContainer">
              {t("BB-ShiftPreference-shiftPattern-error-text", "Select at least one option.")}
            </InputFooter>
          )}
        </Col>
      </Col>
      <Col padding={{ top: "S400" }}>
        <Button
          variant={ButtonVariant.Primary}
          style={{ width: "100%" }}
          onClick={handleSavePreference}
          id="saveShiftPreference"
        >
          {t("BB-shift-preference-save-button", "Save Preferences")}
        </Button>
      </Col>
    </Col>
  );
};

export const mapStateToProps = (state: MapStateToProps) => {
  return state;
};

export default connect(mapStateToProps)(ApplicationShiftPreferences);