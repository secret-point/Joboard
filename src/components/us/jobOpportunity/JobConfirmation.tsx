import React, { useEffect } from 'react';
import { Col, Row } from "@amzn/stencil-react-components/layout";
import JobConfirmationCard from "../../common/jobOpportunity/JobConfirmationCard";
import { H4, Text } from '@amzn/stencil-react-components/text';
import { Button, ButtonVariant } from "@amzn/stencil-react-components/button";
import { translate as t } from "../../../utils/translator";
import { connect } from "react-redux";
import { ScheduleState } from "../../../reducers/schedule.reducer";
import { useLocation } from "react-router";
import {
    getPageNameFromPath,
    parseQueryParamsArrayToSingleItem,
    resetIsPageMetricsUpdated
} from "../../../helpers/utils";
import { addMetricForPageLoad } from "../../../actions/AdobeActions/adobeActions";
import { CommonColors } from "../../../utils/colors";
import { IconArrowLeft, IconSize } from "@amzn/stencil-react-components/icons";
import { checkAndBoundGetApplication, getLocale, handleSubmitJobConfirmation, routeToAppPageWithPath } from "../../../utils/helper";
import { PAGE_ROUTES } from "../../pageRoutes";
import { boundGetScheduleDetail } from "../../../actions/ScheduleActions/boundScheduleActions";
import queryString from "query-string";
import { ApplicationState } from '../../../reducers/application.reducer';
import { JobState } from '../../../reducers/job.reducer';
import { boundGetJobDetail } from '../../../actions/JobActions/boundJobDetailActions';
import { uiState } from '../../../reducers/ui.reducer';
import { CandidateState } from "../../../reducers/candidate.reducer";
import { boundGetCandidateInfo } from "../../../actions/CandidateActions/boundCandidateActions";
import {boundResetBannerMessage} from "../../../actions/UiActions/boundUi";
import DebouncedButton from '../../common/DebouncedButton';

interface MapStateToProps {
    application: ApplicationState
    schedule: ScheduleState
    job: JobState
    ui: uiState,
    candidate: CandidateState
}

export const JobConfirmation = ( props: MapStateToProps ) => {

    const { schedule, application, job, ui, candidate } = props;
    const isLoading = ui.isLoading;
    const { search, pathname } = useLocation();
    const pageName = getPageNameFromPath(pathname);
    const queryParams = parseQueryParamsArrayToSingleItem(queryString.parse(search));
    const scheduleDetail = schedule.results.scheduleDetail;
    const { scheduleId, applicationId, jobId } = queryParams;
    const jobDetail = job.results;
    const applicationDetail = application.results;
    const { JOB_OPPORTUNITIES } = PAGE_ROUTES;
    const candidateData = candidate.results.candidateData;

    useEffect(() => {
        boundGetCandidateInfo();
    },[])

    useEffect(() => {
        jobId && jobId !== jobDetail?.jobId && boundGetJobDetail({ jobId: jobId, locale: getLocale() })
    }, [jobDetail, jobId]);

    useEffect(() => {
        checkAndBoundGetApplication(applicationId);
    }, [applicationId]);

    useEffect(() => {
        scheduleId && boundGetScheduleDetail({
            locale: getLocale(),
            scheduleId: scheduleId
        })
    }, [scheduleId]);

    useEffect(() => {
        jobDetail && scheduleDetail && applicationDetail && candidateData && addMetricForPageLoad(pageName);

    }, [jobDetail, scheduleDetail, applicationDetail, candidateData]);

    useEffect(() => {
        return () => {
            //reset this so as it can emit new pageload event after being unmounted.
            resetIsPageMetricsUpdated(pageName);
        }
    },[])

    const handleConfirmJob = () => {
        boundResetBannerMessage();
        if(applicationDetail && scheduleDetail && jobDetail){
            handleSubmitJobConfirmation(applicationDetail, jobDetail, scheduleDetail);
        }
    }

    return (
        <Col gridGap={10}>
            <Row
                gridGap={5}
                alignItems="center"
                width="fit-content"
                color={CommonColors.Blue70}
                padding='S200'
                onClick={() => {
                    routeToAppPageWithPath(JOB_OPPORTUNITIES);
                }}
                style={{ cursor: 'pointer' }}
            >
                <IconArrowLeft size={IconSize.ExtraSmall}/>
                <Text fontSize="T100" fontWeight="medium">
                    {
                        t('BB-JobOpportunity-back-to-indexPage-link', 'View Jobs')
                    }
                </Text>
            </Row>
            <Col className="jobConfirmationHeader" gridGap={8} padding={{ top: 'S300' }}>
                <H4>{scheduleDetail?.externalJobTitle}</H4>
                <Text fontSize='T200'>{scheduleDetail?.briefJobDescription}</Text>
            </Col>
            {scheduleDetail && <JobConfirmationCard schedule={scheduleDetail}/>}
            <Col className="selectJobButtonContainer" padding='S300'>
                <DebouncedButton
                    disabled = { !applicationDetail || !scheduleDetail || !jobDetail || isLoading}
                    onClick={handleConfirmJob}
                    variant={ButtonVariant.Primary}
                >
                    {t('BB-JobOpportunity-select-job-button', 'Select this job')}
                </DebouncedButton>
            </Col>
        </Col>
    )
}

const mapStateToProps = ( state: MapStateToProps ) => {
    return state;
};

export default connect(mapStateToProps)(JobConfirmation);
