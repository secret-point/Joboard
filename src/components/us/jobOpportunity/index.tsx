import React, { useEffect } from "react";
import { Col, Row } from "@amzn/stencil-react-components/layout";
import JobOpportunity from "./JobOpportunity";
import JobConfirmation from "./JobConfirmation";
import { IconArrowLeft, IconSize } from "@amzn/stencil-react-components/icons";
import { Text } from "@amzn/stencil-react-components/text";
import { JOB_OPPORTUNITY_PAGE } from "../../../utils/enums/common";
import { CommonColors } from "../../../utils/colors";
import JobDescription from "./JobDescription";
import { translate as t } from "../../../utils/translator";
import { connect } from "react-redux";
import { uiState } from "../../../reducers/ui.reducer";
import { boundSetJobOpportunityPage } from "../../../actions/UiActions/boundUi";

interface MapStateToProps {
    ui: uiState
}

const JobOpportunityContainer = (props: MapStateToProps) => {
    const { ui } = props;
    const { jobOpportunity } = ui;
    const { currentPage } = jobOpportunity;
    const { INDEX, JOB_CONFIRMATION, JOB_DESCRIPTION } = JOB_OPPORTUNITY_PAGE;

    useEffect(() => {
        window.scroll(0, 0);
    })

    const renderPage = ( pageType: JOB_OPPORTUNITY_PAGE ) => {
        switch(pageType) {
            case INDEX:
                return <JobOpportunity/>;

            case JOB_CONFIRMATION:
                return <JobConfirmation />;

            case JOB_DESCRIPTION:
                return <JobDescription/>;

            default:
                return <JobOpportunity />
        }
    }


    const handleBack = () => {
        if(currentPage === JOB_CONFIRMATION) {
            boundSetJobOpportunityPage(INDEX);
        }
        else {
            boundSetJobOpportunityPage(JOB_CONFIRMATION);
        }
    }

    return (
        <Col>
            {
                currentPage !== INDEX &&
                <Row
                    gridGap={5}
                    alignItems="center"
                    width="fit-content"
                    color={CommonColors.Blue70}
                    padding='S200'
                    onClick={handleBack}
                    style={{ cursor: 'pointer' }}
                >
                    <IconArrowLeft size={IconSize.ExtraSmall}/>
                    <Text fontSize="T100" fontWeight="medium">
                        {currentPage === JOB_CONFIRMATION ?
                            t('BB-JobOpportunity-back-to-indexPage-link', 'View Jobs') :
                            t('BB-JobOpportunity-back-to-jobConfirmation-link','Back')
                        }
                    </Text>
                </Row>
            }
            {
                renderPage(currentPage)
            }
        </Col>
    )
}

const mapStateToProps = ( state: MapStateToProps ) => {
    return state;
};

export default connect(mapStateToProps)(JobOpportunityContainer)
