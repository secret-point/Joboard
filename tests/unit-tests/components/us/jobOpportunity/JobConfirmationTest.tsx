import React from "react";
import { shallow } from "enzyme";
import { useLocation } from 'react-router-dom';
import { JobConfirmation } from "../../../../../src/components/us/jobOpportunity/JobConfirmation";
import { TestInitUiState, TEST_APPLICATION_ID, TEST_APPLICATION_STATE, TEST_CANDIDATE_STATE, TEST_JOB_ID, TEST_JOB_STATE, TEST_SCHEDULE_ID, TEST_SCHEDULE_STATE } from "../../../../test-utils/test-data";

describe("JobConfirmation", () => {
  const mockLocation = {
    pathname: "/job-opportunities/job-confirmation",
    search: `?jobId=${TEST_JOB_ID}&applicationId=${TEST_APPLICATION_ID}&scheduleId=${TEST_SCHEDULE_ID}`,
    hash: '',
    state: null
  };
  const mockUseLocation = useLocation as jest.Mock;
  mockUseLocation.mockReturnValue(mockLocation);

  it("should match snapshot", () => {
    const shallowWrapper = shallow(
      <JobConfirmation
        candidate={TEST_CANDIDATE_STATE}
        job={TEST_JOB_STATE}
        application={TEST_APPLICATION_STATE}
        schedule={TEST_SCHEDULE_STATE}
        ui={TestInitUiState}
      />);

    expect(shallowWrapper).toMatchSnapshot();
  });
});
