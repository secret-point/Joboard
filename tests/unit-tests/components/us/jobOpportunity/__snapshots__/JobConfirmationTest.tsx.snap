// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JobConfirmation should match snapshot 1`] = `
<Col
  gridGap={10}
>
  <Row
    alignItems="center"
    color="#1768C9"
    gridGap={5}
    onClick={[Function]}
    padding="S200"
    style={
      Object {
        "cursor": "pointer",
      }
    }
    width="fit-content"
  >
    <IconArrowLeft
      display="block"
      size={16}
    />
    <Text
      font="primary"
      fontSize="T100"
      fontWeight="medium"
      hyphens="none"
    >
      View Jobs
    </Text>
  </Row>
  <Col
    className="jobConfirmationHeader"
    gridGap={8}
    padding={
      Object {
        "top": "S300",
      }
    }
  >
    <H4
      color="neutral90"
      font="primary"
      fontSize="T400"
      fontWeight="regular"
      hyphens="none"
      margin={0}
    />
    <Text
      font="primary"
      fontSize="T200"
      hyphens="none"
    />
  </Col>
  <JobConfirmationCard
    schedule={
      Object {
        "scheduleId": "test-schedule-id",
      }
    }
  />
  <Col
    className="selectJobButtonContainer"
    padding="S300"
  >
    <DebouncedButton
      disabled={false}
      onClick={[Function]}
      variant="primary"
    >
      Select this job
    </DebouncedButton>
  </Col>
</Col>
`;
